internal/migrations/020-digest-actions.sql

```diff
@@ -1,231 +1,617 @@
         
-        $preserved_count := $preserved_count + 1;
+        $valid_stream_refs := array_append($valid_stream_refs, $valid_candidate.stream_ref);
+        $valid_day_indexes := array_append($valid_day_indexes, $valid_candidate.day_index);
     }
     
-    -- Insert HIGH marker if different from OPEN and CLOSE
-    if ($high_time != $open_time OR $high_created_at != $open_created_at) AND ($high_time != $close_time OR $high_created_at != $close_created_at) {
-        $high_type := 2;
-        if $low_time = $high_time AND $low_val = $high_val AND $low_created_at = $high_created_at {
-            $high_type := $high_type + 4;  -- HIGH is also LOW
-        }
-        
-        INSERT INTO primitive_event_type (stream_ref, event_time, type)
-        VALUES ($stream_ref, $high_time, $high_type);
+    -- Step 2: TRUE BULK OHLC Processing using UNNEST for ALL streams simultaneously
+    if array_length($valid_stream_refs) > 0 {
+        -- Build arrays for bulk processing with UNNEST
+        $ohlc_stream_refs INT[] := ARRAY[]::INT[];
+        $ohlc_day_starts INT[] := ARRAY[]::INT[];
+        $ohlc_day_ends INT[] := ARRAY[]::INT[];
+        $ohlc_open_times INT[] := ARRAY[]::INT[];
+        $ohlc_open_created_ats INT[] := ARRAY[]::INT[];
+        $ohlc_close_times INT[] := ARRAY[]::INT[];
+        $ohlc_close_created_ats INT[] := ARRAY[]::INT[];
+        $ohlc_high_times INT[] := ARRAY[]::INT[];
+        $ohlc_high_created_ats INT[] := ARRAY[]::INT[];
+        $ohlc_low_times INT[] := ARRAY[]::INT[];
+        $ohlc_low_created_ats INT[] := ARRAY[]::INT[];
         
-        $preserved_count := $preserved_count + 1;
-    }
-    
-    -- Insert LOW marker if different from OPEN, HIGH, and CLOSE
-    if ($low_time != $open_time OR $low_created_at != $open_created_at) AND ($low_time != $close_time OR $low_created_at != $close_created_at) AND ($low_time != $high_time OR $low_created_at != $high_created_at) {
-        INSERT INTO primitive_event_type (stream_ref, event_time, type)
-        VALUES ($stream_ref, $low_time, 4);
+        -- Step 2a: Bulk OHLC calculation for ALL streams using single query with UNNEST
+        for $bulk_ohlc in
+        WITH RECURSIVE stream_days AS (
+            -- Use UNNEST to create table of all stream/day combinations
+            SELECT u.stream_ref, u.day_index,
+                   (u.day_index * 86400) as day_start,
+                   ((u.day_index * 86400) + 86400) as day_end
+            FROM UNNEST($valid_stream_refs, $valid_day_indexes) AS u(stream_ref, day_index)
+        ),
+        stream_record_counts AS (
+            -- Count records for each stream in single query
+            SELECT sd.stream_ref, sd.day_index, sd.day_start, sd.day_end,
+                   COUNT(pe.event_time) as record_count
+            FROM stream_days sd
+            LEFT JOIN primitive_events pe ON pe.stream_ref = sd.stream_ref
+                AND pe.event_time >= sd.day_start AND pe.event_time <= sd.day_end

```

Comment:
hey @MicBun, to solve this very issue, isn't better to define the day as being not inclusive between the two intervals?

as in:

| day | day_start | day_end |
| 1   | 0         | 86399   |
| 2   | 86400     | 172799  |

then the query would be:

```suggestion
                AND pe.event_time >= sd.day_start AND pe.event_time < sd.day_end
```

my concern here is about having the same value on both dates, or conditioned to having conflict or not to be part of one day or the other

or did I get this issue wrong?

Reply:
How should we define day_start and day_end? What's our strategy for handling conflicting values?

While using `<` offers faster implementation, consider this edge case scenario:

**Stream A (hourly ingestion):**
```
82800 | 80
86400 | 100
```
With hourly ingestion, it appears the day closes at 80 and opens the next day at 100.

**However, in real-time scenarios:**
```
82800 | 80
...
86399 | 99.99
86400 | 100
```
This presents a specific edge case where the apparent day-end value differs significantly between batch and real-time processing.

The resolution depends on how we define day_end boundaries. Would it be more efficient to establish clear day_end definitions upfront rather than adding conditional logic to handle date conflicts? Anyway, I'm fine with changing it to `<` instead of `<=` if we can just ignore this edge case. What approach do you think would be most robust? @outerlook